edge/pkg/edgehub/task/taskexecutor/node_rollback.go

```diff
@@ -0,0 +1,56 @@
+package taskexecutor
+
+import (
+	"fmt"
+	"os/exec"
+
+	"k8s.io/klog/v2"
+
+	commontypes "github.com/kubeedge/kubeedge/common/types"
+	api "github.com/kubeedge/kubeedge/pkg/apis/fsm/v1alpha1"
+	"github.com/kubeedge/kubeedge/pkg/util/fsm"
+	"github.com/kubeedge/kubeedge/pkg/version"
+)
+
+func rollbackNode(taskReq commontypes.NodeTaskRequest) (event fsm.Event) {
+	event = fsm.Event{
+		Type:   "Rollback",
+		Action: api.ActionSuccess,
+	}
+	var err error
+	defer func() {
+		if err != nil {
+			event.Action = api.ActionFailure
+			event.ErrorMsg = err.Error()
+		}
+	}()
+
+	var upgradeReq *commontypes.NodeUpgradeJobRequest
+	upgradeReq, err = getTaskRequest(taskReq)
+	if err != nil {
+		return
+	}
+
+	err = rollback(upgradeReq)
+	if err != nil {
+		return
+	}
+	return event
+}
+
+func rollback(upgradeReq *commontypes.NodeUpgradeJobRequest) error {
+	klog.Infof("Begin to run rollback command")
+	upgradeCmd := fmt.Sprintf("keadm rollback edge --upgradeID %s --historyVersion %s >> /tmp/keadm.log 2>&1",
```

Comment:
I don't see upgradeID and historyVersion flag in keadm rollback?

Reply:
I've modified it. The ability to rollback after a failure in real time is not presently available. Because it need a more complex design, such as failure plans.